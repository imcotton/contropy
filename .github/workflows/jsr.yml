name: Publish to JSR



on:
  push:
    tags:
      - v/*



jobs:

  publish-JSR:

    runs-on: ubuntu-latest
    timeout-minutes: 3

    permissions:
      contents: read
      id-token: write

    steps:

      - name: Setup repo
        uses: actions/checkout@08c6903cd8c0fde910a37f88322edcfb5dd907a8 # v5.0.0

      - name: Setup deno
        uses: denoland/setup-deno@e95548e56dfa95d4e1a28d6f422fafe75c4c26fb # v2.0.3
        with:
          deno-version: v2

      - run: deno publish

