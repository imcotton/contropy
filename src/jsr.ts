import {
    sha256,
    http_get,
    encode_hex as to_hex,
} from './utils.ts';





export function gen (pkg: string, ver: string, {

        agent = 'contropy' as string,
        signal = AbortSignal?.timeout(5_000),
        endpoint = 'https://jsr.io',

} = {}): Promise<Response> {

    const url = `${ endpoint }/${ pkg }/${ ver }_meta.json`;

    return http_get(url, agent, signal);

}





export type Result = string;





export function jsr (

        pkg: string,
        ver: string,
        endpoint?: string,

): Promise<Result> {

    return jsr_task(() => gen(pkg, ver, { endpoint }));

}





export async function jsr_task (

        task: () => Promise<Response>,

): Promise<Result> {

    const hash = await task()
        .then(res => res.arrayBuffer())
        .then(sha256)
        .then(to_hex)
    ;

    return hash;

}

