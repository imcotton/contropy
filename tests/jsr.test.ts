import c, { agent } from './common.ts';

import { jsr_task as task, gen } from '../src/jsr.ts';





c.describe('jsr', function () {

    const pkgs = [

        [ '@std/bytes', '1.0.6' ],
        [ '@std/encoding', '1.0.10' ],

        [ '@hono/hono', '4.9.0' ],
        [ '@hono/standard-validator', '0.1.4' ],

        [ '@scure/base', '1.2.6' ],
        [ '@noble/ed25519', '2.2.3' ],

    ] satisfies Array<[ string, string ]>;

    for (const [ pkg, ver ] of pkgs) {

        c.it_on_ci_and(`has pkg ${ pkg }@${ ver }`, async function (t) {

            const [ signal, abort ] = c.use_signal_abort_controller();

            const res = await task(() => gen(pkg, ver, { agent, signal }));

            await c.assertSnapshot(t, res);

            abort();

        });

    }

});

