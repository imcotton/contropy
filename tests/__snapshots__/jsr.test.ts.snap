export const snapshot = {};

snapshot[`jsr > has pkg @std/bytes@1.0.6 1`] = `"f6ac6adbd8ccd99314045f5703e23af0a68d7f7e58364b47d2c7f408aeb5820a"`;

snapshot[`jsr > has pkg @std/encoding@1.0.10 1`] = `"8783c6384a2d13abd5e9e87a7ae0520a30e9f56aeeaa3bdf910a3eaaf5c811a1"`;

snapshot[`jsr > has pkg @hono/hono@4.9.0 1`] = `"49df671bca96af6ab7899ea5ec1f4776b43c4cf864e472affe847bed3bb6b9a1"`;

snapshot[`jsr > has pkg @hono/standard-validator@0.1.4 1`] = `"91dec5fa213b549f9b4215ec4314b3caae23691258070ee436e32e553ea44b79"`;

snapshot[`jsr > has pkg @scure/base@1.2.6 1`] = `"24f115a886d67bfb869ade2c1ada041bfe7f69c677c9ee3725d1ed38ad2d4ac1"`;

snapshot[`jsr > has pkg @noble/ed25519@2.2.3 1`] = `"77d984ee6b823c8b1a4a36aa56105d60e696a4e7b599b821712d1251481e2f9c"`;
